### PR TITLE
optimize `superset` method of `IntervalSet`

### PR DESCRIPTION
**Pull request by SparrowLii**
Given that intervals in the `IntervalSet` are sorted and strictly separated( it means the `end` of the previous interval will not be equal to the `start` of the next interval), we can reduce the complexity of the `superset` method from O(NMlogN) to O(2N) (N is the number of intervals and M is the length of each interval)

### CODE REVIEW

**Comment by lcnr** [comment]
impl looks good

While you're at it, can you document the invariants of `IntervalSet`? Your impl requires that adjacent intervals are merged. That is okay as e.g. `[1..=4, 5..=7]` is currently not a possible `IntervalSet`.

I would like to have that documented, or even better, I would like a method `fn check_invariants` which just asserts that and is used after mutating operations if `debug_assertions` are enabled.

@bors try @rust-timer queue

**Comment by lcnr** [comment]
nice to see that checking the invariants actually found a bug :laughing: even if that bug probably doesn't result in bad behavior as it only affects empty domains.

@bors r+

File: compiler/rustc_index/src/interval.rs [hunk 1]
```diff
@@ -186,8 +218,21 @@ impl<I: Idx> IntervalSet<I> {
...
+    // Check the intervals are valid, sorted and non-adjacent
+    fn check_invariants(&self) -> bool {
+        let mut current: Option<u32> = None;
+        for (start, end) in &self.map {
+            if start > end || current.map_or(false, |x| x + 1 >= *start) {
+                return false;
+            }
+            current = Some(*end);
+        }
+        current.map_or(true, |x| x < self.domain as u32)
```
[the remark this answers is not in the document]
Comment (SparrowLii) [reply compiler/rustc_index/src/interval.rs:234]: I think we'd better make sure all intervals don't exceed `domain`, after all, both `insert_all` and `insert_range` use domain as the range maximum. I increased the `domain` in the test from 3000 to 10000, which I think should be reasonable

**Pull request merged**